### ISSUE TITLE
Add a Parameterized RobolectricGradleTestRunner?

### ISSUE THREAD

**Issue by hidroh** [robolectric/robolectric#1934]
Hi, with the [robolectric-gradle-plugin](https://github.com/robolectric/robolectric-gradle-plugin) now deprecated, `RobolectricGradleTestRunner` seems to be the only option to run tests in CLI. However, for parameterized tests, currently we only have [`ParameterizedRobolectricTestRunner`](https://github.com/robolectric/robolectric/blob/master/robolectric/src/main/java/org/robolectric/ParameterizedRobolectricTestRunner.java), which internally uses `RobolectricTestRunner`. This causes `ResourcesNotFoundException` to parameterized tests when I try to remove robolectric gradle plugin and run them.

A very quick and simple fix would be to make a copy of `ParameterizedRobolectricTestRunner` and simply swap `RobolectricTestRunner` with `RobolectricGradleTestRunner`, but may be the runner can be injected or annotated in some ways?

**Comment by jaredsburrows** [comment]
+1

**Comment by erd** [comment]
See #871.

**Comment by WonderCsabo** [comment]
@erd i think you incorrectly closed this bug, this is not a duplicate.

**Comment by erd** [comment]
@WonderCsabo Isn't this issue and #871 both about getting the parameterized runner working? If not, what's the difference?

**Comment by WonderCsabo** [comment]
These are two separate issues. This is about the parameterized runner uses the old resources paths, not the Gradle one, and there is no easy way to use the parameterized runner with gradle.

The other issue is unrelated. That is about classloading problems: the runner parameter objects are created using the normal class loader, but during the test the Robolectric classlaoder is used. This means if you have Android objects (say a Parcelable) in you parameters, you cannot use it in the test, and you will have reflective exceptions if you try to do. Please not i opened a PR already to resolve that problem.

**Comment by erd** [comment]
Ok - Thanks for the explanation!

**Comment by WonderCsabo** [comment]
No problem at all!

**Comment by jongerrish** [comment]
Can't we just make RobolectricTestRunner support parameters? Then it will work for any subclasses.

**Comment by 1zaman** [comment]
> Can't we just make RobolectricTestRunner support parameters? Then it will work for any subclasses.

That would be a good solution, though it would need quite a bit of conditional logic. I recently [extended][parameterized extension] my own Robolectric test runner to work with `Parameterized` (via the `UseParametersRunnerFactory` annotation), and it required duplicating a lot of code from `RobolectricTestRunner` in order to prevent the `@Before` and `@After` methods from being executed.

[parameterized extension]: https://github.com/edx/edx-app-android/blob/4dd11a542785d2689b5b4936124502954ec12239/VideoLocker/src/test/java/org/edx/mobile/CustomRobolectricTestRunnerWithParameters.java

**Comment by 1zaman** [comment]
Note that the existing `ParameterizedRobolectricTestRunner` seems to be quite broken, apart from being based on (a subset of) the implementation in JUnit 4.11 (which was not pluggable). Robolectric itself seems to be using JUnit 4.8.2 as a dependency, which was released back in 2010 :open_mouth:

**Comment by rspeyer** [comment]
Is there any update on this issue? I would be interested in this functionality as well

**Comment by jaredsburrows** [comment]
@hidroh Can you make a PR with your own here: https://github.com/hidroh/materialistic/blob/534bc19db4e479736758aed8cb243cb63da1c7bb/app/src/test/java/io/github/hidroh/materialistic/test/ParameterizedRobolectricGradleTestRunner.java?

**Comment by paleozogt** [comment]
+1

**Comment by michaelbanholzer** [comment]
+1

**Comment by trevjonez** [comment]
Would be nice if we could get first party support from the robolectric team to implement an API like https://github.com/square/burst

I spent some time trying to wrap robolectric with burst but it quickly turns into a reflective mess due to private visibility. Once I had constructor params working I moved on method params and got stuck due to multiple class loaders making reflection fail. Too far out of my realm of understanding.
